### DISCUSSION TITLE
Wanting to roam between 2 APs with the same SSID

### DISCUSSION THREAD

**Discussion by jimruxton** [pschatzmann/arduino-audio-tools#1092] [Q&A] [answered]
I'm using an ESP32S3 module that is grabbing an Opus stream via UDP. I'm trying to grab this stream off a WiFi mesh network. This network has 2 APs with the same SSID. I am getting a lot of glitches as if the device keeps   trying to grab both APs. This happens even if the ESP32 is right beside one of the APs. I found a version of the Arduino libraries that integrates roaming for ESP32s. Wondering if there is a way to integrate this into my code which uses audio tools . Below is my current code.

  /**
 * @file example-serial-receive.ino
 * @author Phil Schatzmann
 * @brief Receiving audio via IP and writing to I2S
 * @version 0.1
 * @date 2022-03-09
 *
 * @copyright Copyright (c) 2022
 */

#include "AudioTools.h"
//#include "AudioLibs/Communication.h"
#include "Communication/UDPStream.h" //used for latest download but revert
#include "AudioCodecs/CodecOpus.h"
#include <driver/adc.

const char *ssid = "NETGEAR41";
const char *password = "xxxxxxxxx";
//UDPStream stream;
I2SStream out;
UDPStream stream(ssid, password);
const int udpPort = 8000;
EncodedAudioStream dec(&out, new OpusAudioDecoder());
StreamCopy copier(dec, stream);
AudioInfo info(16000, 1, 16);
void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

   Serial.println("starting I2S...");
  auto cfgi = out.defaultConfig(TX_MODE);
  cfgi.copyFrom(info);
  out.begin(cfgi);
 
  cfgi.buffer_size = 1024; 
  cfgi.buffer_count = 20; 
 
  out.begin(cfgi);
  stream.begin(8000);
  dec.begin(cfgi);

    dec.setNotifyAudioChange(out);

  }
void loop() {
     copier.copy();
}

**Comment by jimruxton** [comment]
This is the repository with the roaming enabled. Wondering if it's possible to integrate this into my program above.https://github.com/Roobuck-RnD/arduino-esp32-roaming/

**Reply by pschatzmann** [reply, marked as answer]
Where do you see any challenge?

**Comment by jimruxton** [comment]
It seems to me that audio tools has all the wifi stuff embedded in it and isn't using the arduino WiFi.h for example. I figured I would somehow have to take stuff from the roaming library and integrate it into the audio tools source tree?

**Reply by pschatzmann** [reply]
Not at all: just do the login in your sketch

**Comment by jimruxton** [comment]
Sorry,
Just so I'm clear. So you are saying I can just replace the traditional
Arduino core with the one in the link above and add your library to that
and the program should use the roaming version?
Thanks

**Comment by pschatzmann** [comment]
Yes, and I am also saying that if you do not like the __optional login__ functionality, you are free to login into the Wifi by using your login logic in your sketch.

**Comment by jimruxton** [comment]
Great, thank you so much for clarifying this.
